Java/core/src/main/java/com/amazon/randomcutforest/genericstore/RCStore.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.genericstore;
+
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+/**
+ * Reference counted store.
+ * 
+ * @param <T>
+ */
+public class RCStore<T> extends Store<T> {
```

Comment:
Should we keep the reference counting a separate concept and not extend the Store<T> class?

This could expose the following APIs
public void incrementRefCount(int index)
public void decrementRefCount(int index)
public int getRefCount(int index)

The PointStore can extend both the Store<T> class and RCStore class

Reply:
Java doesn't support extending more than one class. I think the best we could do along these lines would be to make a class like ReferenceCounter which implements those methods, and then give the PointStore a ReferenceCounter instance. Between the two approaches (defining a separate class or keeping the current subclass approach), I don't have a strong opinion.

Keep in mind that the generic store classes are somewhat exploratory: I wanted to see if taking a more standard Java approach would be a big enough improvement over the column-oriented array based approach that it would be worth paying the price in terms of additional memory used. Currently, I feel that it's not enough of an improvement and I think we'll end up deleting these classes from the package.

Reply:
Keeping the details of reference counting in a separate class would be better conceptually. But not a blocker as of now.